### PR TITLE
rust-analyzer: 2021-08-02 -> 2021-08-09 and some fixes for the extension

### PR DESCRIPTION
**Pull request by oxalica**
###### Motivation for this change
`rust-analyzer`:
- Bump version

`vscode-extension.matklad.rust-analyzer`:
- Use individual extension version instead of "0.4.0-dev" forever. The extension version is from upstream released vsix file and is the same as the one published on vscode market place.
- Apply some patches to extension's `package.json`, which follows [upstream dist script](https://github.com/rust-analyzer/rust-analyzer/blob/41949748a6123fd6061eb984a47f4fe780525e63/xtask/src/dist.rs#L39-L65).

Tested to work in vscodium.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 133290` at c825b1b1 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built successfully:</summary>
  <ul>
    <li>rust-analyzer</li>
    <li>rust-analyzer-unwrapped</li>
    <li>vscode-extensions.matklad.rust-analyzer</li>
  </ul>
</details>
<details>
  <summary>4 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`buildPhase` should probably contain `runHook preBuild` and `runHook postBuild`.

Near pkgs/misc/vscode-extensions/vscode-utils.nix:21:12:
```
   |
21 |     inherit configurePhase buildPhase dontPatchELF dontStrip;
   |            ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

unzip is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/misc/vscode-extensions/vscode-utils.nix:25:5:
```
   |
25 |     buildInputs = [ unzip ] ++ buildInputs;
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`configurePhase` should probably contain `runHook preConfigure` and `runHook postConfigure`.

Near pkgs/misc/vscode-extensions/vscode-utils.nix:21:12:
```
   |
21 |     inherit configurePhase buildPhase dontPatchELF dontStrip;
   |            ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/name-and-version.md">name-and-version</a>

Did you mean to pass `pname` instead of `name` to `mkDerivation`?

Near pkgs/misc/vscode-extensions/vscode-utils.nix:18:5:
```
   |
18 |     name = "vscode-extension-${name}";
   |     ^
```

Near pkgs/misc/vscode-extensions/rust-analyzer/default.nix:45:10:
```
   |
45 |   inherit version vsix;
   |          ^
```
</li>
  </ul>
</details>

File: pkgs/misc/vscode-extensions/rust-analyzer/default.nix [hunk 1]
```diff
@@ -12,23 +12,34 @@ let
...
+    # Follows https://github.com/rust-analyzer/rust-analyzer/blob/41949748a6123fd6061eb984a47f4fe780525e63/xtask/src/dist.rs#L39-L65
     postInstall = ''
-      npm run package
-      mkdir $vsix
-      cp ${pname}.vsix $vsix/${pname}.zip
+      jq '
+        .releaseTag = $ENV.releaseTag |
+        .enableProposedApi = false |
+        walk(del(.["$generated-start"]?) | del(.["$generated-end"]?))
+      ' package.json >package.json.new
```
Comment (happysalada) [inline pkgs/misc/vscode-extensions/rust-analyzer/default.nix:32]: sponge from moreutils was made just for that.
If you add `moreutils` to nativeBuildInputs, you can change this to 
`' package.json | sponge package.json`
no more need to create a new file and mv it.

**Comment by happysalada** [comment]
Result of `nixpkgs-review pr 133290` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>rust-analyzer</li>
    <li>rust-analyzer-unwrapped</li>
    <li>vscode-extensions.matklad.rust-analyzer</li>
  </ul>
</details>

**Comment by happysalada** [comment]
Ran update.sh and didn't have any problems.
I'm happy to test it at a later date to verify, but I don't think it's necessary.

Let me know what you want to do with the `moreutils`, otherwise fine by me to merge

File: pkgs/misc/vscode-extensions/rust-analyzer/default.nix (hunk 1 above)
Comment (oxalica) [reply pkgs/misc/vscode-extensions/rust-analyzer/default.nix]: FIxed.

**Comment by oxalica** [comment]
Forget to commit line `.version = $ENV.version` before. Now it's fixed and rebased.

**Comment by happysalada** [comment]
Just giving it one more round of nixpkgs-review for the sake of it.

**Comment by happysalada** [comment]
Result of `nixpkgs-review pr 133290` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>rust-analyzer</li>
    <li>rust-analyzer-unwrapped</li>
    <li>vscode-extensions.matklad.rust-analyzer</li>
  </ul>
</details>

**Comment by happysalada** [comment]
I don't see anything else.

Thank you for the PR!

**Pull request merged**